### ISSUE TITLE
EHN: Loading heavy data set

### ISSUE THREAD

**Issue by pburdet** [hyperspy/hyperspy#592]
@to266 edit, 2015-09-25:

There is an increasingly grave need to be able to load datasets bigger than the available (RAM) memory. Unfortunately since various different formats have different functionalities, it has to be done on a format-by-format basis, and I guess we all agree that the first target should be the HyperSpy native HDF5.

I suggest using [`dask`](https://github.com/blaze/dask) for the numerical routines (it looks to be the perfect choice by supporting multiple formats).
### The landmarks for HDF5:
- [x] Read-only of big datasets (i.e. everyone can open and plot (with limited navigators) big signals) (#735)
- [x] Write big datasets (in addition to opening and plotting, you can stack various other signals (that individually fit in memory) and save the result as one big HDF5) (#736)
- [x] functions for big data (most notably [`dask`](https://github.com/blaze/dask) ) (used to think `signal.map`, but probably not worth it). `dask` has to catch up or we have to implement a bunch of things ourselves
- [x] enable batchPCA and batchICA (could not find the routines quickly, should be in either `scipy` or `scikit-learn` hopefully
- [ ] [optionally] rewrite (now hopefully deprecated) convenience functions (such as `fold`) that were mainly required for PCA / other machine learning
# Original @pburdet message:

I'm trying to load 2500 images as one 3D image (one image can be found under this link)

https://app.box.com/s/h73aoypja66n1fe0q2bl

The total size on the disk is 12.5GB. I'm able to open them as one stack with Imagej, but not with hyperspy. I'm running out of RAM even with 32GB. I tried `mmap=True` and `stack=False`. Nothing work.

Imagej load them as 8-bits images, Hyperspy as 16-bits with 3 channels (as if RGB). Can we control the way hyperspy load the data? Does the mmap is working correctly?
## TODO
- [ ] Enhance `mmap` loading of stacks: currently it loads all the files in memory before stacking them in disk (!). This is probably easy to fix and would solve the main problem pointed out by @pburdet.
- [ ] When loading multiple files at once, add the ability to change the dtype after loading each individual file.
- [ ] Implement `mmap` for all readers that can support it.
- [ ] Option of not saving the orginial_metadata (usefull when many files are loaded)
- [ ] Improve docstring regarding `load` `mmap` argument.

**Comment by francisco-dlp** [comment]
Good point. The `mmap` functionality in HyperSpy is very limited. In particular, for loading it only works with the ripple format. I fully agree that this is something that we should prioritize. I will add this to the wishlist but, if anybody feels like working on this for HyperSpy 0.9, feel free to (request a) change of milestone. I will also draft a list of tasks in the issue header.

**Comment by pburdet** [comment]
Do you know why I cannot open an object of which the size on the disk is more than two time smaller than my available memory?

**Comment by pburdet** [comment]
mmap is an important functionality for sure

**Comment by francisco-dlp** [comment]
> Do you know why I cannot open an object of which the size on the disk is more than two time smaller than my available memory?

Can you provide more details?

**Comment by francisco-dlp** [comment]
(I have just updated the header with the TODO list)

**Comment by pburdet** [comment]
I have a folder containing 2500 images, as the one under the link. The total size is 12GB. I cannot load them with hyperspy

**Comment by francisco-dlp** [comment]
For some reason I couldn't download the file from the link, could you send it using a different service?

One simple guess: are the images stored in a compressed format?

Things to check: can you load them all with `stack=False`

If `stack=True` and `mmap=False` then you'll actually need double the size in memory to perform the task. With `mmap=True` the operation should require exactly the size in memory of all the files as `stack` will store the stack in a temporary file used as a memmap. Ideally, stack should remove from memory each individual image after storing it in the file but this is not how it is working at present. Did any of my guesses hit the problem?

**Comment by magnunor** [comment]
For HDF5 we could utilize the chunking functionality, where it is possible to seamlessly load certain parts of the HDF5-file: https://www.hdfgroup.org/HDF5/doc/Advanced/Chunking/ . This would make it a little bit slower, since one would have to read from disk quite often, but would make it actually possible to load the files.

HDF5 also has some pretty nice in-built compressions, which makes it useful for storing lots of data. 

I've also been discussing this with the people over at http://emdatasets.lbl.gov/ , which are grappling with the same issues: datasets much larger than commonly available amounts of memory (RAM). 

One "trick" for loading very large datafiles is using a SSD as swap under Linux, which will probably shorten the lifetime of the SSD with quite a bit. But it will probably be cheaper than buying large amounts of memory.

**Comment by vidartf** [comment]
> using a SSD as swap

@magnunor : In `load()` just specify `mmap=True` and `mmap_dir` to a folder on your SSD and you'll use that folder as scratch disk. This sidesteps memory / swap, and relies on numpy for effective routines for working with the data set (I hope and think that simple operations don't try to read the entire file into memory to calculate).

**Comment by vidartf** [comment]
> In particular, for loading it only works with the ripple format.

@francisco-dlp I'm not sure if I follow. Currently memmap is only used for stacks, but is successfully applied regardless of the source data format. Or dam I misunderstanding something?

**Comment by pburdet** [comment]
Thanks @francisco-dlp, I tried the different option you suggested. But none are working. It could be a problem of image compression or a problem of hyperspy transforming the data type of the image. Can you have a look at the image?

https://documents.epfl.ch/users/p/pb/pburdet/www/cropped_slice_00067_z=0.4643um.tif

**Comment by pburdet** [comment]
The TODO list corresponds well to my problem. We might also have the option of not saving the orignal_metadata when ```stack=True```, as it is known to slow things down.

**Comment by francisco-dlp** [comment]
@vidartf, you are half right: currently memmap is used for stacks  and it is successfully applied regardless of the source data format. However, some formats (currently only ripple) provide the option to load the data using numpy's memmap function. I think that it'll be nice to extend this feature to other formats. @magnunor mentioned hdf5 and, indeed, I agree that this should be our first target.

@pburdet, regarding the dtype issue, I think that it is a bug (or a feature) in @cgohlke 's tifffile module which we distribute with HyperSpy to support scientific tiff files. I have tested two other modules to read tifffiles with your file: `scipy.misc.imread` and [imread](https://github.com/luispedro/imread). Scipy's imread reads your image fine while luispedro's imread behaves as @cgohlke's one. It seems that ImageJ has its own undocumented tiff format (see luispedro/imread#19) and this may (or may not) be the problem. I think that it'll be worth to report the issue to @cgohlke and the imread developers, would you like to do that @pburdet?

In the meantime you can use the following workaround to use scipy's imread instead of tifffile to load your image:

```python

>>> import hyperspy.hspy as hs
>>> import hyperspy.io_plugins.tiff
>>> import hyperspy.io_plugins.image
>>> for i in range(len(hyperspy.io_plugins.tiff.file_extensions)):
...         hyperspy.io_plugins.image.file_extensions.append(hyperspy.io_plugins.tiff.file_extensions.pop())
...     
>>> 
>>> im = load("cropped_slice_00067_z=0.4643um.tif")
>>> im.data.dtype
dtype('uint8')

```

**Comment by magnunor** [comment]
I'm very interested in having this capability for HDF5, so I could probably join the work on implementing it. But, unfortunately, I won't have time to work on this until November - December.

**Comment by vidartf** [comment]
@francisco-dlp Ah, I see now. I also see there is an argument to the ripple plugin called `mmap_mode` to control this behavior. If we're implementing this further, I guess the docs could need an update too.

**Comment by francisco-dlp** [comment]
I fully agree—I've added it to the TODO list.

**Comment by cgohlke** [comment]
What is the dtype issue? The `cropped_slice_00067_z=0.4643um.tif` file has a `PhotometricInterpretation = Palette color` tag. TIFF palettes are stored as `(3, 256) uint16`, hence the image with applied palette is `(3, length, width) uint16`.

**Comment by francisco-dlp** [comment]
@cgohlke, the issue is that ImageJ opens this image as (length, width) uint8 and @pburdet expected tifffile to do the same. I fully agree with your arguments. However, this file's color_map has some striking peculiarities:

* cm[:,2:] = range(2, 256)
* All values are multiples of 256 except for  cm[0, 0] = 37920 = 148.125, cm[1, 0] = 37392 =  146.0625 * 256 and cm[:, 1] = 1304 = 256 * 5.09375
* cm[0, 1:] = cm[1, 1:] = cm[2, 1:]

(color_map  (cm) dimensions are (3, 256) )

So, it looks like a gray palette with the exception of (0, 0, 0) and (1, 1, 1). Interestingly, the 0 index value is only used for the top border of the image. It seems that scipy.misc.imread simply doesn't apply the palette and hence the top border of the image ends up filled with zeros. ImageJ reads the image as gray uint8 but, interestingly, even though the top border has also a value of 0, it displays it as if it had a value of cm[2, 0] = 36464 / 256 = 144.

May this simply be a convention that uses the palette to set the border gray level to cm[2, 0] (144 in this case) and mark it as background, possibly to make it easy to crop the border if needed? I couldn't make any sense of the value of cm[:, 1], which should be 256 instead of 1304. Actually ImageJ gives them a value of 1—maybe it is used to mark positions in the image?

**Comment by pburdet** [comment]
The image is acquired with a proprietary software (not imagej). They are several saving options. ```cropped_slice_00067_z=0.4643um.tif``` was saved with option 8-bits and pallette (I don't understand how this option could be usefull on machine that only do gray scale image). It might be just the tiff format that is not well applied by the manufacturer. I'll test the different saving option next time I'm on the microscope

**Comment by to266** [comment]
(please see my edit of the main issue as well)

#### Work copy
The one somewhat significant side-effect of out-of-memory processing (so starting with 3rd landmark, maybe even the writing one as well) is that the temporary results will be stored in the disk. Since no one would like to change a file just by opening it, it means we would need to create "working copy" of the file that would most likely be just as big as the original (assuming most of the weight is in the data itself, not metadata). 

There are a couple of questions of different importance for this topic as well:
 - do we want it to be explicit - i.e. allow the user to load the data knowing that the file will be changed if any processing takes place, and that he is responsible for having a backup?
 - If not explicit, do we create the new file only when required, or upon loading the signal?
 - Do we store a complete HyperSpy HDF5 as the working copy, or just the data? We could make use of HDF5 references and links if the latter.

**Comment by vidartf** [comment]
> Work copy

I think we could open the file as a work copy by default, but have an option to open "in place". For work copies, it is common to have a preference where one can set the "scratch disk", defaulting to e.g. temp.dir.

There is also a "middle-ground" option, where changes will be written to the original file, but all operations write "atomically". This basically means writing out all changes to disk somewhere temporary, and then once the operation finishes successfully, flush those changes to the original file. For maximum safety (but highest disk cost), you could have a work-copy + atomic writes.

**Comment by francisco-dlp** [comment]
>  - do we want it to be explicit - i.e. allow the user to load the data knowing that the file will be changed if any processing takes place, and that he is responsible for having a backup?

I agree with @vidartf, +1   on this as an (non default) option.

> - If not explicit, do we create the new file only when required, or upon loading the signal?

+1 on creating it when required.

Having a default scratch disk location sounds good to me as long as  i) we can guarantee that the file is deleted when deleting the signal or ending the session ii) we inform the user that the file has been stored in the tmp place and will be deleted unless he does something about it iii) we provide a method to move the file from tmp to the user specified location for when the user decides to keep the file.

> - Do we store a complete HyperSpy HDF5 as the working copy, or just the data? We could make use of HDF5 references and links if the latter.

+1 on storing a complete HyperSpy hdf5 copy.

**Comment by to266** [comment]
...ok, the `dask` is suitable for us as a replacement for HDF5 dataset, as it does not have `__setitem__` methods, hence can only change / replace value by performing calculations. I suppose that then the approach would be to (in the out-of-memory case) always keep the reference to the hdf5 dataset, and construct `dask.array` as required, overwriting it in the hdf5 afterwards.

I've change the main issue accordingly

**Comment by vidartf** [comment]
BTW: Would `dask` help with lazily evaluating compressed data?

For example: Say you have an enormous spectrum image that you process chunk-wise from disk, and then perform PCA/ICA compression. Most likely, the compressed data should fit in memory, and as such might be preferable to work for reasons of speed. However, to work with it, we would need to be able to lazily evaluate (uncompress) the data at each navigation index. In terms of computing power, such linear combinations of components should be unproblematic, but a design decision would have to be made at when and where the uncompression should happen. As such, `Signal.data` is the most general place to do it. There are several packages out there that allows such numpy-compatible lazy evalution, but we should make sure that we pick a package that does all the things we want (including the things in this issue), as the different packages are unlikely to be internally compatible/combineable.

PS: It would also be nice if we were able to save PCA/ICA compressed signals ("decomposition models") in their compressed form, both for storage and transfer of data.

**Comment by francisco-dlp** [comment]
If I understand you well, you're mainly proposing a lazy version of `get_decomposition_model` and `get_bss_model`. I think that it is not only a nice idea, but also a necessity when dealing with big data sets. Why not opening a different issue about it?

Saving PCA/ICA results "in their compressed form" is currently possible using `Signal.learning_results.crop_decomposition_dimension` and `Signal.learning_results.save`. However, again I agree that we should reimagine the way in which we store the decomposition and BSS results. But again, I think that this deserves an issue of its own.

**Comment by vidartf** [comment]
@francisco-dlp Yes, it definitely needs an issue on its own, but for now my priority was on making sure the that `dask` is the best option of the different packages out there. Basically trying to make our decision as "future-proof" as possible.

**Comment by to266** [comment]
@vidartf as far as I can tell from [this](https://github.com/blaze/dask#related-work) description, it's the tool for us. 
Having quickly looked at the `get_decompostion_model` (which is, more generally, `_calculate_recmatrix`), it should be trivial to implement with `dask`, as the dot product is already there.
Also the computations can be done lazily (as much as possible with given chunk size) as well, and the arrays support iteration as required, hence we may be able to just replace the `_calculate_recmatrix` with a generator in some cases.

The other `dask` advantage is that we can, if we decide to, go one optimization level "higher" and use the `blaze` it's been built for. 

In short, looks like an incredibly awesome library/package in my eyes, love it already :)

**Comment by to266** [comment]
@pburdet, https://github.com/hyperspy/hyperspy/pull/736 should fix your original problem (loading and stacking images)

**Comment by to266** [comment]
Having had a better look at `signal.map` I think it's not worth it trying to rewrite most of our functions to use it - not only it's made for a different purpose (to simplify iteration over navigation / axes passing __for the user__, as we mix them up when compared to standard numpy), but it would have to become even more complex (forcing certain axes passing, for example), and yet still it would not be good enough, as a good chunk of numpy functions are not present in dask (though there are equivalent, sometimes even better things).

There is also the annoying bit that in currently released version `dask.Array` has much less functions than there are available in their github (e.g. `reshape` is missing, which we use extensively). So we have to either push them to release quicker, delay this branch development until they do, or implement / copy the code ourselves.... And none of these options look good.

**Comment by tjof2** [comment]
How much of this was addressed in #1219, and can this be closed?

**Comment by to266** [comment]
I believe so!
